sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -1431,6 +1415,67 @@ void Scheduler::GraphBuilder::cancelFusion(QueueImplPtr Queue,
   PlaceholderCmd->setFusionStatus(KernelFusionCommand::FusionStatus::CANCELLED);
 }
 
+static bool isPartOfFusion(Command *Cmd, KernelFusionCommand *Fusion) {
+  if (Cmd->getType() == Command::RUN_CG) {
+    return static_cast<ExecCGCommand *>(Cmd)->MFusionCmd == Fusion;
+  }
+  return false;
+}
+
+static bool checkForCircularDependency(Command *, bool, KernelFusionCommand *);
+
+static bool createsCircularDependency(Command *Cmd, bool PredPartOfFusion,
+                                      KernelFusionCommand *Fusion) {
+  if (isPartOfFusion(Cmd, Fusion)) {
+    if (PredPartOfFusion) {
+      // If this is part of the fusion and the predecessor also was, we can stop
+      // the traversal here. A direct dependency between two kernels in the same
+      // fusion will never form a cyclic dependency and by iterating over all
+      // commands in a fusion, we will detect any cycles originating from the
+      // current command.
+      return false;
+    } else {
+      // If the predecessor was not part of the fusion, but the current command
+      // is, we have found a potential cycle in the dependency graph.
+      return true;
+    }
+  }
+  return checkForCircularDependency(Cmd, false, Fusion);
+}
+
+static bool checkForCircularDependency(Command *Cmd, bool IsPartOfFusion,
+                                       KernelFusionCommand *Fusion) {
+  // Check the requirement dependencies.
+  for (auto &Dep : Cmd->MDeps) {
+    auto *DepCmd = Dep.MDepCommand;
+    if (!DepCmd) {
+      continue;
+    }
+    if (createsCircularDependency(DepCmd, IsPartOfFusion, Fusion)) {
+      return true;
+    }
```

Comment:
```suggestion
    return createsCircularDependency(DepCmd, IsPartOfFusion, Fusion);
```
The same goes for the code below

Reply:
I'm not sure I completely understand your suggestion here, sorry. 

The current code only returns if `createsCircularDependency(...) == true`, otherwise it will process the next dependency or move on to the next loop. 

With your suggestion, the first loop would always return from `checkForCircularDependency` on the first iteration (where `DepCmd != nullptr`), so it would change the semantics of the code. 

Did I misunderstand your suggestion?

Reply:
You're totally right, I somehow managed to misinterpret that, never mind.